### PR TITLE
Add ENUM class for country codes

### LINKED ISSUE #7886: Introduce country constants
### Description

Introduce constants for country code values and replace all hardcoded values. 

### Acceptance criteria

All references to country codes should be handled using PHP constants.

### Additional context

paJDYF-bFG-p2

### PR DESCRIPTION
**Pull request by zmaglica**
Fixes #7886 

#### Changes proposed in this Pull Request
This pull request introduces a class that functions similarly to an enum class, specifically designed for managing country codes.
#### Testing instructions

* Make sure that all tests are passing

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7925` or branch name `dev/7886-add-country-codes-enum` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=dev/7886-add-country-codes-enum) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: a8e9f93d83add60323e44d5fadf6b965351dd5d0
- Build time: 2024-01-19 15:19:24 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/constants/class-country-codes.php [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Country Codes constants
+ *
+ * @psalm-immutable
+ */
+class Country_Codes extends Base_Constant {
```
Comment (jessy-p) [inline includes/constants/class-country-codes.php]: Can we use `Country` instead of `Country_Codes`?
It would be easier to read e.g. `Country::JAPAN === $account_country`

Comment (zmaglica) [reply includes/constants/class-country-codes.php]: I've considered your point, but there's a potential for confusion. Some might expect `Country::JAPAN` to return 'Japan' or its equivalent in the currently configured language. To avoid this confusion, I opted for `Country_Code`.

Comment (jessy-p) [reply includes/constants/class-country-codes.php]: Makes sense, so I guess we can use `Country_Code`?

Comment (naman03malhotra) [reply includes/constants/class-country-codes.php]: `Country_Code` makes more sense, +1

File: includes/constants/class-country-codes.php [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+	const AFGHANISTAN                      = 'AF';
+	const ALAND_ISLANDS                    = 'AX';
```
Comment (naman03malhotra) [inline includes/constants/class-country-codes.php]: Can we write a test for the output of this class? If that test breaks devs will know that some change has been made here willingly or not

**Comment by naman03malhotra** [comment]
@zmaglica do you plan to replace country code in `locale-info.php` as well?

**Comment by zmaglica** [comment]
> @zmaglica do you plan to replace country code in `locale-info.php` as well?

The linked file, is taken from the WooCommerce Repository ([link](https://github.com/woocommerce/woocommerce/blob/d9fd8cee20780bed1596da796b116b6148a5205e/plugins/woocommerce/i18n/currency-info.php#L4)), which serves as a fallback for older WooCommerce installations lacking locale-info.php at that time. Given its introduction three years ago and its specific use-case, I excluded it from the country codes refactor, since all our supported versions of WooCommerce uses one provided by WooCommerce plugin.

File: includes/constants/class-country-codes.php (hunk 1 above)
Comment (zmaglica) [reply includes/constants/class-country-codes.php]: Good point. Refactored in https://github.com/Automattic/woocommerce-payments/pull/7925/commits/d3ce810bdb73032b918a81dd53cb501f245ff213

File: includes/constants/class-country-codes.php (hunk 2 above)
Comment (zmaglica) [reply includes/constants/class-country-codes.php]: Added tests here: https://github.com/Automattic/woocommerce-payments/pull/7925/commits/09424655aee6221a74b62a1e6dcbe209fe081734

**Comment by naman03malhotra** [comment]
>all our supported versions of WooCommerce uses one provided by WooCommerce plugin.

Can you please link the file where this logic is present?

**Comment by naman03malhotra** [comment]
> Testing instructions
Make sure that all tests are passing

@zmaglica is it enough to verify If this change is working or do you think we need some manual testing as well?

**Comment by zmaglica** [comment]
> > all our supported versions of WooCommerce uses one provided by WooCommerce plugin.
> 
> Can you please link the file where this logic is present?

There are 2 places where this logic can be found and only in these two places, the mentioned `locale-info.php` is used.

One of the places is located here: https://github.com/Automattic/woocommerce-payments/blob/6610e65aabe084bc45b66c2994385c0cc62a4d2b/includes/class-wc-payments-localization-service.php#L113-L118

If you check the code, you will notice that the `locale-info.php` is loaded from the WooCommerce plugin folder, and if that file is missing, the fallback `locale-info.php` is loaded from our codebase

Another instance where this particular file is loaded can be found here: https://github.com/Automattic/woocommerce-payments/blob/6610e65aabe084bc45b66c2994385c0cc62a4d2b/includes/admin/class-wc-payments-admin.php#L795-L799

Basically, this code check is WooCommerce version below 6.0, and if it is, it loads `locale-info.php` from WooPayments plugin folder, instead of the WooCommerce plugin folder.

And if you take a look inside the WooCommerce repository, the locale-info.php is added in [5.7.0 version](https://github.com/woocommerce/woocommerce/blob/937b6d66c80dfae5e6ecbdf929b9ead0894af22c/plugins/woocommerce/i18n/locale-info.php#L4)

In terms of content, these file from WooCommerce and WooPayments are identical.

**Comment by zmaglica** [comment]
> > Testing instructions
> > Make sure that all tests are passing
> 
> @zmaglica is it enough to verify If this change is working or do you think we need some manual testing as well?

Unit tests cover this topic very well, so we don't have to do manual testing. This is a static class, and if it not included or used somewhere, the unit tests will fail.

**Review by jessy-p — APPROVED** [review]
LGTM!
